tests/run-tests-parallel.py

```diff
@@ -0,0 +1,110 @@
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+# STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+# THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+# The parallelism celing can be controlled with MONERO_PARALLEL_TEST_JOBS env variable,
+# for example:
+#
+# MONERO_PARALLEL_TEST_JOBS=1     tests/run-tests-parallel.py
+# MONERO_PARALLEL_TEST_JOBS=nproc tests/run-tests-parallel.py
+#
+# The minimum between the ceiling and a currently selected reasonable number of threads is used in the end.
+# The reasonable number is selected as a number, that delivers the solution in the shortest time.
+"""
+
+import os
+import sys
+import subprocess
+import multiprocessing
+from timeit import default_timer as timer
+
+NUM_PROC_MAX_REASONABLE = 2 # This might be increased, once the core_tests are divided into many more independent pieces or simply sped up
+TESTS_EXCLUDED_REGEX = "unit_tests|functional_tests_rpc" # These tests collide with core_tests
+ERR_CODE = 1
+
+def get_forced_job_ceiling():
+    try:
+        ceiling_str = os.environ['MONERO_PARALLEL_TEST_JOBS']
+    except KeyError:
+        ceiling = multiprocessing.cpu_count()
+        print("No parallelism ceiling selected. Defaulting to nproc, so", ceiling)
+    else:
+        if ceiling_str == "nproc":
+            ceiling = multiprocessing.cpu_count()
+        else:
+            ceiling = int(ceiling_str)
+
+        print("Parallelism ceiling selected. Using", ceiling, "jobs.")
+
+    return ceiling
+
+
+def run(num_proc_final, fail_fast=False):
+    """
+    Run the excluded tests first, then everything but excluded.
+    In the current situation, the excluded tests are the most probable and quickest to fail,
+    giving an early feedback when fail_fast is set to true.
+    """
+    cmds = []
+    cmds.append(get_ctest_command(num_proc_final, '-R'))
+    cmds.append(get_ctest_command(num_proc_final, '-E'))
+    
+    error = False
+    for cmd in cmds:
+        status = run_cmd(cmd)
+        if status != 0:
+            error = True
+            if fail_fast:
+                return status
+
+    return ERR_CODE if error else 0
+
+def get_ctest_command(num_proc_final, option):
+    cmd = ['ctest', '-j{}'.format(num_proc_final), option, TESTS_EXCLUDED_REGEX]
+    return cmd
+
+def run_cmd(cmd):
+    result = subprocess.run(cmd, stderr=sys.stderr, stdout=sys.stdout)
+    return result.returncode
+
+
+job_ceiling = get_forced_job_ceiling()
+num_proc_final = min(NUM_PROC_MAX_REASONABLE, job_ceiling)
+print("Job number ceiling is" ,job_ceiling, "and the reasonable maximum is currently", NUM_PROC_MAX_REASONABLE, ".")
+print("Using the minimum of the two, so", num_proc_final, "jobs.")
+
+start = timer()
+#TODO: set fail_fast=True after mining / RPC random test failures are resolved
+ret = run(num_proc_final, fail_fast=False)
```

Comment:
@selsta : Since the mining test ("Functional") is now stable, do you think I could set the fail fast to True?

Reply:
The above test will be fixed with: https://github.com/monero-project/monero/pull/7646

Reply:
I don't like the concept of fail fast personally. I want to see everything that fails.

Reply:
Then I will remove that TODO from line 104.